runtime/js/40_http.js

```diff
@@ -0,0 +1,203 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+"use strict";
+
+((window) => {
+  const { Request, dontValidateUrl, fastBody, Response } =
+    window.__bootstrap.fetch;
+  const { Headers } = window.__bootstrap.headers;
+  const errors = window.__bootstrap.errors.errors;
+  const core = window.Deno.core;
+  const { ReadableStream } = window.__bootstrap.streams;
+
+  function flatEntries(obj) {
+    const entries = [];
+    for (const key in obj) {
+      entries.push(key);
+      entries.push(obj[key]);
+    }
+    return entries;
+  }
+
+  function startHttp(conn) {
+    const rid = Deno.core.jsonOpSync("op_http_start", conn.rid);
+    return new HttpConn(rid);
+  }
+
+  class HttpConn {
+    #rid = 0;
+
+    constructor(rid) {
+      this.#rid = rid;
+    }
+
+    get rid() {
+      return this.#rid;
+    }
+
+    close() {
+      core.close(this.#rid);
+    }
+
+    async next() {
+      try {
+        const [
+          connectionClosed,
+          requestBodyRid,
+          responseSenderRid,
+          method,
+          headersList,
+          url,
+        ] = await Deno.core.jsonOpAsync("op_http_request_next", this.#rid);
```

Comment:
```suggestion
        const {
          0: connectionClosed,
          1: requestBodyRid,
          2: responseSenderRid,
          3: method,
          4: headersList,
          5: url,
        } = await Deno.core.jsonOpAsync("op_http_request_next", this.#rid);
```

Reply:
Is there any advantage to using this syntax? Destructuring this array doesn't seem to have performance hit according to flamegraphs

Reply:
keyed lookups are much faster and cheaper than creating iterator and collecting it
flamegraph won't likely show diff in this case as it's very small change compared to everything else

but the diff can be clearly seen in bytecode


keyed: `const { 0: zero: 1: one } = ()`
```
[generated bytecode for function: aaae (0x0c85eeda8e89 <SharedFunctionInfo aaae>)]
Parameter count 1
Register count 4
Frame size 32
   77 S> 0xc85eeda96d6 @    0 : 1b 02             LdaImmutableCurrentContextSlot [2]
         0xc85eeda96d8 @    2 : ab 00             ThrowReferenceErrorIfHole [0]
         0xc85eeda96da @    4 : 26 f9             Star r2
   60 S> 0xc85eeda96dc @    6 : 0b                LdaZero 
         0xc85eeda96dd @    7 : 26 f8             Star r3
         0xc85eeda96df @    9 : 2a f9 00          LdaKeyedProperty r2, [0]
         0xc85eeda96e2 @   12 : 26 fb             Star r0
   69 S> 0xc85eeda96e4 @   14 : 0c 01             LdaSmi [1]
         0xc85eeda96e6 @   16 : 26 f8             Star r3
         0xc85eeda96e8 @   18 : 2a f9 02          LdaKeyedProperty r2, [2]
         0xc85eeda96eb @   21 : 26 fa             Star r1
         0xc85eeda96ed @   23 : 0d                LdaUndefined 
   84 S> 0xc85eeda96ee @   24 : aa                Return 
Constant pool (size = 1)
0xc85eeda9689: [FixedArray] in OldSpace
 - map: 0x2bc831201721 <Map>
 - length: 1
           0: 0x2b156b67c129 <String[5]: #array>
Handler Table (size = 0)
Source Position Table (size = 12)
0x0c85eeda96f1 <ByteArray[12]>
```

iterator: `const [zero, one] = ()`
```
[generated bytecode for function: aaae (0x35e99e468e89 <SharedFunctionInfo aaae>)]
Parameter count 1
Register count 13
Frame size 104
   69 S> 0x35e99e469716 @    0 : 1b 02             LdaImmutableCurrentContextSlot [2]
         0x35e99e469718 @    2 : ab 00             ThrowReferenceErrorIfHole [0]
         0x35e99e46971a @    4 : 26 f9             Star r2
   69 E> 0x35e99e46971c @    6 : b1 f9 00 02       GetIterator r2, [0], [2]
         0x35e99e469720 @   10 : a0 07             JumpIfJSReceiver [7] (0x35e99e469727 @ 17)
         0x35e99e469722 @   12 : 61 b8 00 fb 00    CallRuntime [ThrowSymbolIteratorInvalid], r0-r0
         0x35e99e469727 @   17 : 26 f7             Star r4
         0x35e99e469729 @   19 : 28 f7 01 04       LdaNamedProperty r4, [1], [4]
         0x35e99e46972d @   23 : 26 f8             Star r3
         0x35e99e46972f @   25 : 11                LdaFalse 
         0x35e99e469730 @   26 : 26 f6             Star r5
         0x35e99e469732 @   28 : 27 ff f3          Mov <context>, r8
   56 S> 0x35e99e469735 @   31 : 25 f6             Ldar r5
         0x35e99e469737 @   33 : 97 25             JumpIfToBooleanTrue [37] (0x35e99e46975c @ 70)
         0x35e99e469739 @   35 : 10                LdaTrue 
         0x35e99e46973a @   36 : 26 f6             Star r5
         0x35e99e46973c @   38 : 58 f8 f7 0a       CallProperty0 r3, r4, [10]
         0x35e99e469740 @   42 : 26 f2             Star r9
         0x35e99e469742 @   44 : a0 07             JumpIfJSReceiver [7] (0x35e99e469749 @ 51)
         0x35e99e469744 @   46 : 61 b0 00 f2 01    CallRuntime [ThrowIteratorResultNotAnObject], r9-r9
         0x35e99e469749 @   51 : 28 f2 02 08       LdaNamedProperty r9, [2], [8]
         0x35e99e46974d @   55 : 97 0f             JumpIfToBooleanTrue [15] (0x35e99e46975c @ 70)
         0x35e99e46974f @   57 : 28 f2 03 06       LdaNamedProperty r9, [3], [6]
         0x35e99e469753 @   61 : 26 f2             Star r9
         0x35e99e469755 @   63 : 11                LdaFalse 
         0x35e99e469756 @   64 : 26 f6             Star r5
         0x35e99e469758 @   66 : 25 f2             Ldar r9
         0x35e99e46975a @   68 : 8b 03             Jump [3] (0x35e99e46975d @ 71)
         0x35e99e46975c @   70 : 0d                LdaUndefined 
         0x35e99e46975d @   71 : 26 fb             Star r0
   62 S> 0x35e99e46975f @   73 : 25 f6             Ldar r5
         0x35e99e469761 @   75 : 97 25             JumpIfToBooleanTrue [37] (0x35e99e469786 @ 112)
         0x35e99e469763 @   77 : 10                LdaTrue 
         0x35e99e469764 @   78 : 26 f6             Star r5
         0x35e99e469766 @   80 : 58 f8 f7 0c       CallProperty0 r3, r4, [12]
         0x35e99e46976a @   84 : 26 f2             Star r9
         0x35e99e46976c @   86 : a0 07             JumpIfJSReceiver [7] (0x35e99e469773 @ 93)
         0x35e99e46976e @   88 : 61 b0 00 f2 01    CallRuntime [ThrowIteratorResultNotAnObject], r9-r9
         0x35e99e469773 @   93 : 28 f2 02 08       LdaNamedProperty r9, [2], [8]
         0x35e99e469777 @   97 : 97 0f             JumpIfToBooleanTrue [15] (0x35e99e469786 @ 112)
         0x35e99e469779 @   99 : 28 f2 03 06       LdaNamedProperty r9, [3], [6]
         0x35e99e46977d @  103 : 26 f2             Star r9
         0x35e99e46977f @  105 : 11                LdaFalse 
         0x35e99e469780 @  106 : 26 f6             Star r5
         0x35e99e469782 @  108 : 25 f2             Ldar r9
         0x35e99e469784 @  110 : 8b 03             Jump [3] (0x35e99e469787 @ 113)
         0x35e99e469786 @  112 : 0d                LdaUndefined 
         0x35e99e469787 @  113 : 26 fa             Star r1
         0x35e99e469789 @  115 : 0c ff             LdaSmi [-1]
         0x35e99e46978b @  117 : 26 f4             Star r7
         0x35e99e46978d @  119 : 26 f5             Star r6
         0x35e99e46978f @  121 : 8b 07             Jump [7] (0x35e99e469796 @ 128)
         0x35e99e469791 @  123 : 26 f4             Star r7
         0x35e99e469793 @  125 : 0b                LdaZero 
         0x35e99e469794 @  126 : 26 f5             Star r6
         0x35e99e469796 @  128 : 0f                LdaTheHole 
         0x35e99e469797 @  129 : a7                SetPendingMessage 
         0x35e99e469798 @  130 : 26 f3             Star r8
         0x35e99e46979a @  132 : 25 f6             Ldar r5
         0x35e99e46979c @  134 : 97 26             JumpIfToBooleanTrue [38] (0x35e99e4697c2 @ 172)
         0x35e99e46979e @  136 : 27 ff f1          Mov <context>, r10
         0x35e99e4697a1 @  139 : 28 f7 04 0e       LdaNamedProperty r4, [4], [14]
         0x35e99e4697a5 @  143 : 9f 1d             JumpIfUndefinedOrNull [29] (0x35e99e4697c2 @ 172)
         0x35e99e4697a7 @  145 : 26 f0             Star r11
         0x35e99e4697a9 @  147 : 58 f0 f7 10       CallProperty0 r11, r4, [16]
         0x35e99e4697ad @  151 : a0 15             JumpIfJSReceiver [21] (0x35e99e4697c2 @ 172)
         0x35e99e4697af @  153 : 26 ef             Star r12
         0x35e99e4697b1 @  155 : 61 b0 00 ef 01    CallRuntime [ThrowIteratorResultNotAnObject], r12-r12
         0x35e99e4697b6 @  160 : 8b 0c             Jump [12] (0x35e99e4697c2 @ 172)
         0x35e99e4697b8 @  162 : 26 f1             Star r10
         0x35e99e4697ba @  164 : 0b                LdaZero 
         0x35e99e4697bb @  165 : 6d f5             TestReferenceEqual r6
         0x35e99e4697bd @  167 : 99 05             JumpIfTrue [5] (0x35e99e4697c2 @ 172)
         0x35e99e4697bf @  169 : 25 f1             Ldar r10
         0x35e99e4697c1 @  171 : a9                ReThrow 
         0x35e99e4697c2 @  172 : 25 f3             Ldar r8
         0x35e99e4697c4 @  174 : a7                SetPendingMessage 
         0x35e99e4697c5 @  175 : 0b                LdaZero 
         0x35e99e4697c6 @  176 : 6d f5             TestReferenceEqual r6
         0x35e99e4697c8 @  178 : 9a 05             JumpIfFalse [5] (0x35e99e4697cd @ 183)
         0x35e99e4697ca @  180 : 25 f4             Ldar r7
         0x35e99e4697cc @  182 : a9                ReThrow 
         0x35e99e4697cd @  183 : 0d                LdaUndefined 
   76 S> 0x35e99e4697ce @  184 : aa                Return 
Constant pool (size = 5)
0x35e99e4696a9: [FixedArray] in OldSpace
 - map: 0x37d138201721 <Map>
 - length: 5
           0: 0x241b4b3bc129 <String[5]: #array>
           1: 0x37d138204b71 <String[4]: #next>
           2: 0x37d1382043a9 <String[4]: #done>
           3: 0x37d1382053d1 <String[5]: #value>
           4: 0x37d138204f59 <String[6]: #return>
Handler Table (size = 32)
   from   to       hdlr (prediction,   data)
  (  31, 115)  ->   123 (prediction=0, data=8)
  ( 139, 160)  ->   162 (prediction=0, data=10)
Source Position Table (size = 14)
0x35e99e4697d1 <ByteArray[14]>
```

Reply:
@devcat Once we have this basic thing landed, performance hacks like this will be very welcome. For the moment, we're more concerned about error handling, API, and semantics. Please submit in a follow up. I expect this patch to land today.